drivers/psi5/psi5_nxp_s32.c

```diff
@@ -0,0 +1,519 @@
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/pinctrl.h>
+
+#include <zephyr/drivers/psi5/psi5.h>
+
+#include "Psi5_Ip.h"
+
+struct psi5_nxp_s32_config {
+	uint8_t ctrl_inst;
+	PSI5_Type *base;
+	uint8_t channel_mask;
+	const struct pinctrl_dev_config *pin_cfg;
+	void (*irq_config_func)(void);
+};
+
+struct psi5_nxp_s32_tx_callback {
+	psi5_tx_callback_t callback;
+	void *user_data;
+};
+
+struct psi5_nxp_s32_channel_data {
+	bool started;
+	bool async_mode;
+	struct psi5_nxp_s32_tx_callback tx_callback;
+	struct psi5_rx_callback_configs callback_configs;
+	uint32_t serial_frame_cnt;
+	uint32_t data_frame_cnt;
+	struct k_sem tx_sem;
+	struct k_mutex lock;
+};
+
+struct psi5_nxp_s32_data {
+	struct psi5_nxp_s32_channel_data channel_data[PSI5_CHANNEL_COUNT];
+};
+
+struct psi5_nxp_s32_tx_default_cb_ctx {
+	struct k_sem done;
+	int status;
+};
+
+static int psi5_nxp_s32_start_sync(const struct device *dev, uint8_t channel)
+{
+	const struct psi5_nxp_s32_config *config = dev->config;
+	struct psi5_nxp_s32_data *data = dev->data;
+	struct psi5_nxp_s32_channel_data *channel_data = &data->channel_data[channel];
+	int err;
+
+	if (!(config->channel_mask & BIT(channel))) {
+		return -EINVAL;
+	}
+
+	if (channel_data->started) {
+		return -EALREADY;
+	}
+
+	if (channel_data->async_mode) {
+		return -ENOTSUP;
+	}
+
+	k_mutex_lock(&channel_data->lock, K_FOREVER);
+
+	err = Psi5_Ip_SetChannelSync(config->ctrl_inst, channel, true);
+
+	if (err) {
+		LOG_ERR("Failed to start sync PSI5 %d channel %d", config->ctrl_inst, channel);
+		k_mutex_unlock(&channel_data->lock);
+		return -EIO;
+	}
+
+	channel_data->started = true;
+
+	k_mutex_unlock(&channel_data->lock);
+
+
+	return 0;
+}
+
+static int psi5_nxp_s32_stop_sync(const struct device *dev, uint8_t channel)
+{
+	const struct psi5_nxp_s32_config *config = dev->config;
+	struct psi5_nxp_s32_data *data = dev->data;
+	struct psi5_nxp_s32_channel_data *channel_data = &data->channel_data[channel];
+	int err;
+
+	if (!(config->channel_mask & BIT(channel))) {
+		return -EINVAL;
+	}
+
+	if (!channel_data->started) {
+		return -EALREADY;
+	}
+
+	if (channel_data->async_mode) {
+		return -ENOTSUP;
+	}
+
+	k_mutex_lock(&channel_data->lock, K_FOREVER);
+
+	err = Psi5_Ip_SetChannelSync(config->ctrl_inst, channel, false);
+
+	if (err) {
+		LOG_ERR("Failed to stop sync PSI5 %d channel %d", config->ctrl_inst, channel);
+		k_mutex_unlock(&channel_data->lock);
+		return -EIO;
+	}
+
+	channel_data->started = false;
+
+	k_mutex_unlock(&channel_data->lock);
+
+	return 0;
+}
+
+static int psi5_nxp_s32_do_send(const struct device *dev, uint8_t channel, uint64_t psi5_data,
+				k_timeout_t timeout, psi5_tx_callback_t callback, void *user_data)
+{
+	const struct psi5_nxp_s32_config *config = dev->config;
+	struct psi5_nxp_s32_data *data = dev->data;
+	struct psi5_nxp_s32_channel_data *channel_data = &data->channel_data[channel];
+	int err;
+
+	if (!(config->channel_mask & BIT(channel))) {
```

Comment:
Please move all these condition checks to `psi5_nxp_s32_send`

Reply:
I upadted

Reply:
I updated